File: Depth-First Search/Combinations/Numbers/LET_no77_Combinations.py
class Solution:
    """
    @param n: Given the range of Numbers
    @param k: Given the Numbers of combinations
    @return: All the combinations of k Numbers out of 1..n
    """

    def combine(self, n, k):

        result = []
        self.dfs(n, k, 1, 0, [], result)
        return result

    def dfs(self, n, k, start, used, temp, result):
        if used == k:
            result.append(temp[:])
            return

        for i in range(start, n + 1):
            temp.append(i)
            self.dfs(n, k, i + 1, used + 1, temp, result)
            temp.pop()

File: Depth-First Search/Combinations/Numbers/LET_no77_Combinations.cpp
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> res;
        vector<int> temp;
        helper(n, k, 1, temp, res);
        return res;
    }

    void helper(int n, int k, int start, vector<int>& temp, vector<vector<int>>& res) {
        if (k == 0) {
            res.push_back(temp);
            return;
        }

        for (int i = start; i <= n; i++) {
            temp.push_back(i);
            helper(n, k - 1, i + 1, temp, res);
            temp.pop_back();
        }
    }
};